src/app/query/stacks/bns/bns.utils.ts

```diff
@@ -0,0 +1,157 @@
+import { asciiToBytes } from '@stacks/common';
+import { BnsNamesOwnByAddressResponse } from '@stacks/stacks-blockchain-api-types';
+import {
+  BufferCV,
+  ClarityType,
+  OptionalCV,
+  PrincipalCV,
+  TupleCV,
+  UIntCV,
+  addressToString,
+  bufferCV,
+  cvToHex,
+  deserializeCV,
+  standardPrincipalCV,
+  tupleCV,
+} from '@stacks/transactions';
+
+import { StacksClient } from '@app/query/stacks/stacks-client';
+
+const BNSX_CONTRACT_CONSTS = {
+  contractAddress: 'SP1JTCR202ECC6333N7ZXD7MK7E3ZTEEE1MJ73C60',
+  contractName: 'bnsx-registry',
+} as const;
+
+function bytesToAscii(buffer: Uint8Array) {
+  let ret = '';
+  const end = buffer.length;
+
+  for (let i = 0; i < end; ++i) {
+    ret += String.fromCharCode(buffer[i] & 0x7f);
+  }
+  return ret;
+}
+
+// Fetch an address's "primary name" from the BNSx contract.
+async function fetchBnsxName(client: StacksClient, address: string): Promise<string | null> {
+  try {
+    const addressCV = standardPrincipalCV(address);
+    const addressHex = cvToHex(addressCV);
+    const res = await client.smartContractsApi.callReadOnlyFunction({
+      ...BNSX_CONTRACT_CONSTS,
+      functionName: 'get-primary-name',
+      tip: 'latest',
+      readOnlyFunctionArgs: {
+        sender: address,
+        arguments: [addressHex],
+      },
+    });
+    if (!res.okay || !res.result) return null;
+    const { result } = res;
+    const cv = deserializeCV(result) as OptionalCV<
+      TupleCV<{ name: BufferCV; namespace: BufferCV }>
+    >;
+    if (cv.type === ClarityType.OptionalNone) return null;
+    const { name, namespace } = cv.value.data;
+    const fullName = `${bytesToAscii(name.buffer)}.${bytesToAscii(namespace.buffer)}`;
+    return fullName;
+  } catch (error) {
+    return null;
+  }
+}
+
+// This function is not exported from `@stacks/transactions`
+function principalToString(principal: PrincipalCV): string {
```

Comment:
This is imported in `clarity-value-list.tsx` from `import { principalToString } from '@stacks/transactions/dist/esm/clarity/types/principalCV';` Should either do same or move to a shared function?